lib/modules/versioning/nuget/index.ts

```diff
@@ -1,36 +1,260 @@
-import semver from 'semver';
-import { regEx } from '../../../util/regex';
-import { GenericVersion, GenericVersioningApi } from '../generic';
-import type { VersioningApi } from '../types';
+import type { NewValueConfig, VersioningApi } from '../types';
+import { cmp } from './compare';
+import { parseRange, parseVersion } from './parse';
+import * as nugetRange from './range';
+import type { NugetVersion } from './types';
 
 export const id = 'nuget';
 export const displayName = 'NuGet';
 export const urls = [
   'https://docs.microsoft.com/en-us/nuget/concepts/package-versioning',
 ];
-export const supportsRanges = false;
+export const supportsRanges = true;
+export const supportedRangeStrategies = ['pin', 'replace'];
 
-const pattern = regEx(/^(\d+(?:\.\d+)*)(-[^+]+)?(\+.*)?$/);
+class NugetVersioningApi implements VersioningApi {
+  isCompatible(version: string, _current?: string | undefined): boolean {
+    return this.isValid(version);
+  }
+
+  isSingleVersion(version: string): boolean {
+    if (this.isVersion(version)) {
+      return true;
+    }
+
+    const parsed = parseRange(version);
+    return !!parsed && parsed.type === 'range-exact';
+  }
+
+  isStable(version: string): boolean {
+    const parsedVersion = parseVersion(version);
+    if (parsedVersion) {
+      return !parsedVersion.prerelease;
+    }
+
+    const parsedRange = parseRange(version);
+    if (parsedRange?.type === 'range-exact') {
+      return !parsedRange.version.prerelease;
+    }
+
+    return false;
+  }
+
+  isValid(input: string): boolean {
+    return !!parseVersion(input) || !!parseRange(input);
+  }
+
+  isVersion(input: string | undefined | null): boolean {
+    return input ? !!parseVersion(input) : false;
+  }
+
+  getMajor(version: string): number | null {
```

Comment:
why not using the old base class and reuse as much as possible?

Reply:
☝️

Reply:
Which base class exactly? `GenericVersioningApi` internal representation is not compatible with the whole range of nuget version variations.

Nuget turned out to be one of the most nuanced versioning schemes I encountered, so I think there is no too much opportunity for code reuse, other than coupling with several trivial code lines somewhere else.